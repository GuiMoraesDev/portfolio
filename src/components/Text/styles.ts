import styled, { DefaultTheme } from 'styled-components';

import { TextDefaultPropsThatMakeStyles, VariantProps } from '.';

interface TextProps extends TextDefaultPropsThatMakeStyles {
	tag?: string;
}

const textVariant = (theme: DefaultTheme): Record<VariantProps, string> => ({
	regular: theme.themeColors.text,
	light: theme.themeColors.textLight,
	inverted: theme.themeColors.textInverted,
	black: theme.colors.neutrals[900],
	white: theme.colors.neutrals['050'],
});

export const Text = styled.p.attrs<TextProps>(({ tag }) => ({
	as: tag,
}))<TextProps>`
	position: relative;

	padding: 0;
	margin: 0;

	width: ${({ fullWidth }) => fullWidth && '100%'};

	${({ theme, dimension }) => theme.typography.variants[dimension || 'body2']};

	font-weight: ${({ isBold }) => isBold && 'bold'};

	color: ${({ variant, theme }) => textVariant(theme)[variant || 'regular']};

	text-transform: ${({ isCapitalize }) =>
		isCapitalize ? 'capitalize' : 'none'};

	box-sizing: border-box;
`;
