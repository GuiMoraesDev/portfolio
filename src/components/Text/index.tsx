import React, { HTMLAttributes } from 'react';

import { DefaultTheme } from 'styled-components';

import * as Styles from './styles';

export type VariantProps = 'regular' | 'light' | 'inverted' | "black" | 'white';

export interface TextDefaultPropsThatMakeStyles {
	fullWidth?: boolean;
	dimension?: keyof DefaultTheme['typography']['variants'];
	variant?: VariantProps;
	isBold?: boolean;
	isCapitalize?: boolean;
}

type TagProps = Record<keyof DefaultTheme['typography']['variants'], string>;

interface TextDefaultProps extends TextDefaultPropsThatMakeStyles {
	label: string;
}

type TextProps = TextDefaultProps & HTMLAttributes<HTMLElement>;

const Text = ({
	label,
	variant = 'regular',
	fullWidth = false,
	dimension = 'body2',
	...props
}: TextProps): JSX.Element => {
	const TagsOptions: TagProps = {
		display1: 'strong',
		heading1: 'h1',
		heading2: 'h2',
		heading3: 'h3',
		heading4: 'h4',
		heading5: 'h5',
		body1: 'p',
		body2: 'p',
		body3: 'p',
		body4: 'p',
	};

	const Tag = TagsOptions[dimension];

	return (
		<Styles.Text
			tag={Tag}
			fullWidth={fullWidth}
			dimension={dimension}
			variant={variant}
			{...props}
		>
			{label}
		</Styles.Text>
	);
};

export default Text;
