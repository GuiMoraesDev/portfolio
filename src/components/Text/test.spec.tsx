import React from "react";

import { render } from "tests/test.utils";

import Text from "./index";

describe("Text component", () => {
  it("should render correctly", () => {
    const { container } = render(<Text id="test-Text" label="test-Text" />);

    expect(container.firstChild).toMatchSnapshot();
  });

  it("should render visible", () => {
    const { container } = render(<Text id="test-Text" label="test-Text" />);

    expect(container.firstChild).toBeVisible();
  });
});
