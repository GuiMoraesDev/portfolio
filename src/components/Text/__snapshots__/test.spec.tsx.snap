// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text component should render correctly 1`] = `
.c0 {
  position: relative;
  padding: 0;
  margin: 0;
  font-size: 1.4rem;
  font-weight: 400;
  font-family: 'Roboto',sans-serif;
  line-height: 2.0rem;
  color: #181917;
  text-transform: none;
  box-sizing: border-box;
}

<p
  class="c0"
  id="test-Text"
>
  test-Text
</p>
`;
